### ISSUE TITLE
Support cleaning the cache directory

### ISSUE THREAD

**Issue by ryanccn** [Schniz/fnm#696]
It would be great if `fnm` supported cleaning up the cache directory. Currently, I have **2398 directories** in the cache directory, and I think that having a command (which maybe runs periodically when you run any `fnm` command?)﻿that removes everything in that directory is better than having to do so manually while the folders build up.

**Comment by ryanccn** [comment]
@Schniz

**Comment by Schniz** [comment]
hey there. There is no good solution unfortunately.

1. these are not directories but symlinks. They don't really harm your computer but feel free to remove them: `rm $(dirname $FNM_MULTISHELL_PATH)/*`
2. periodic cleanings aren't gonna cut it 100% because we need to know whether the shell is still in use. however this is not really possible because we can't know for sure which shell have evaluated this env. it was possible if we would say to init fnm with `fnm env --pid=$$` but we don't.
3. as I mentioned time does not tell us anything. We might be able to clean on restarts--but I don't know how to hook into that. If you know how to make this magic happen I'd be happy to pair or get your insights so we can bake this into fnm and clean up the symlinks directory.
4. I have 937 symlinks (shocking, I expected way more). `du $(dirname $FNM_MULTISHELL_PATH)` return 0 bytes, so I am not concerned about it. should I?

**Comment by aesadsklfjsl** [comment]
Maybe this is useful: https://stackoverflow.com/questions/18221348/exit-hook-working-both-on-bash-and-zsh

**Comment by Schniz** [comment]
Thanks for this!!!

**Comment by fengyinxu** [comment]
Hello~ I don't know why fnm needs to create multishells symlinks. Even if you enter the same path again, it still creates a new and different symlinks. I don't understand why it was designed this way.

**Comment by ryanccn** [comment]
> Hello~ I don't know why fnm needs to create multishells symlinks. Even if you enter the same path again, it still creates a new and different symlinks. I don't understand why it was designed this way.

@fengyinxu It's because each shell instance can be using its own version of Node.js, so each multishell symlink is specific to one shell instance.

**Comment by fengyinxu** [comment]
> > Hello~ I don't know why fnm needs to create multishells symlinks. Even if you enter the same path again, it still creates a new and different symlinks. I don't understand why it was designed this way.
> 
> @fengyinxu It's because each shell instance can be using its own version of Node.js, so each multishell symlink is specific to one shell instance.

Typically, what we need is a different version of Nodejs for each project, not each shell instance. I'm not familiar with rust. I'm curious as to why fnm designed it this way (cache, multishells) What is the need for this design. I see there is a PR #865 associated with the issue, but it's a pity that the PR has not been merged now.

**Comment by ryanccn** [comment]
@fengyinxu Commands like `fnm use` change the node version the shell is using; even `.node-version` and `engines.node` support set the node version in the shell where you `cd`ed into the project.

**Comment by MunifTanjim** [comment]
Is there a reason for not creating these symlinks in `$TMPDIR`? 🤔

**Comment by Aalivexy** [comment]
For fish shell user:

```fish
function fnm_clean_up --on-event fish_exit
    rm -r $FNM_MULTISHELL_PATH
end
```

**Comment by Schniz** [comment]
the problem with using hooks is that almost no one exits their shell cleanly, so that hook will not be called. Maybe fish works better than what I tested?

**Comment by Schniz** [comment]
> Is there a reason for not creating these symlinks in $TMPDIR? 🤔 @MunifTanjim 

sorry for replying after more than 1 year. fnm multishells used to be in TMPDIR. Unfortunately, a MacOS change in behavior of tmpdirs made them clean up on hibernation instead of reboot. we had bug reports in this regard, which the new multishell directory fixed. I didn't find any literature about this change but changing the directory fixed it for good, so we left it like that.

I think it's a good tradeoff to have: symlinks are mostly 0-sized and fast, and working software is better than broken software

**Comment by MunifTanjim** [comment]
Is there a way to add a cleanup command? Maybe add the `pid` of the shell as a prefix for `FNM_MULTISHELL_PATH` file. And the cleanup command can remove the file if the process for that `pid` is not running anymore?

Or maybe just add the `pid` prefix part, and no need to add the command. People who wants to cleanup can write their own script?

**Comment by Schniz** [comment]
the problem with pid is that it isn't a shell command but an executable, so I need to use `ppid` for the parent id, but that doesn't cut it either, because what if you have a "setup script" that calls fnm for your shell?

```zsh
# my .zshrc
eval "$(~/.config/setup-fnm)"

# ~/setup
fnm env --use-on-cd
```

then we need to traverse the process tree to understand what's the closest shell

however we can do `fnm env --pid $PID` as an optional argument which will tell "This is the current shell". again it will be problematic when it is composed in another script, but might help cleaning up stuff if necessary

I don't think there's a good and easy solution here. do you have something in mind?

**Comment by Schniz** [comment]
tldr the main concern is that using PID might hurt composition

**Comment by secondetik** [comment]
I don't realize this issue already raised as I created my suggestion on https://github.com/Schniz/fnm/issues/1374

I see that the problem would be on ensuring the clean up command to be executed on any shell exit? 
Using PID sounds complicated so I'm not sure about that.

Also, it makes me wonder why do we need symlink in the first place? Why not assign the path directly to avoid creation of symlink?

**Comment by Cufoon** [comment]
I still can't understand why we should use symbolic link fnm_multishells. I think it's completely unnecessary. Node management tools largely rely on modifying environment variables such as PATH to function. As long as we can put the path of the target Node version into the PATH, we can run the corresponding environment. Even if each terminal has its own corresponding Node version, so what? What's the difference between multiple fnm_multishells pointing to the same actual Node version and directly pointing to an actual Node version?

**Comment by bysiber** [comment]
2398 directories in cache — that's a lot! Can confirm this is still an issue. Node version managers silently accumulate download caches that never get cleaned.

Quick audit on macOS:

```bash
du -sh ~/Library/Caches/fnm_multishells/ 2>/dev/null
du -sh ~/Library/Application\ Support/fnm/ 2>/dev/null
ls ~/Library/Caches/fnm_multishells/ 2>/dev/null | wc -l
```

Manual workaround until a `fnm cache clean` lands:

```bash
# Remove old multishell entries (safe — they're ephemeral per terminal session)
rm -rf ~/Library/Caches/fnm_multishells/*

# Remove unused Node versions
fnm ls
fnm uninstall <version>
```

This is a common pattern across version managers — nvm, Volta, and SDKMAN all have similar open issues requesting built-in cleanup ([nvm#1195](https://github.com/nvm-sh/nvm/issues/1195), [volta#330](https://github.com/volta-cli/volta/issues/330), [sdkman#688](https://github.com/sdkman/sdkman-cli/issues/688)).

For a bird's-eye view of all developer cache paths on macOS (fnm, nvm, Homebrew, Xcode DerivedData, Docker, Gradle, pip, Cargo, and 40+ more), [ClearDisk](https://github.com/bysiber/cleardisk) is a free, open-source menu bar utility that monitors them in real-time.

**Comment by pping** [comment]
```cmd
@echo off
if not defined FNM_ENV (
   set FNM_ENV=true
   set XDG_RUNTIME_DIR=%TEMP%
   for /f "delims=" %%i in ('fnm env --shell cmd') do %%i
)
```

**Comment by futuremotiondev** [comment]
Any official response to this? I just switched over from nvm for windows after doing a bit of research, and while I now love and prefer `fnm` over `nvm` now, this is the only drawback I'm seeing. 

Use case:

I have LOTS of AI mcp servers that spawn a lot of node sessions during the lifecycle of my development process, and the multishells are going to accumulate extremely rapidly due to this. I've put together a little powershell script that cleans out this folder, but this specific issue should really be officially addressed.

Would love an update - will be monitoring this issue moving forward.

**Comment by indykoning** [comment]
Hooking into

> I have 937 symlinks (shocking, I expected way more). du $(dirname $FNM_MULTISHELL_PATH) return 0 bytes, so I am not concerned about it. should I?

because i have run into this before and it's a pain in the ass since you get no warning and suddenly your server is broken.
While symlinks do not take much or any storage space, they DO take inode space. If this is exceeded you will get the same effect as if the disk is at 100% usage (in my case it couldn't even clear the tmp folder during a reboot without recovery mode)

Do note that this high of a usage is dificult to hit, but it may be possible if a ton of shells are opened often.
You can view the inode usage per disk with `df -i` and per folder/file with `du --inodes $(dirname $FNM_MULTISHELL_PATH)`

**Comment by Jeffrey-deng** [comment]
I still can't understand why we need the fnm_multishells symbolic link in every shell. its unnecessary
Why not point directly to the actual version directory? It can also use different versions in various shells.
